tools/jit/gen_jit_dispatch.py

```diff
@@ -0,0 +1,122 @@
+    'std::array<bool, 3>': 'as_bool_array<3>',
+    'Scalar': 'Scalar',
+}
+
+ATTR_ASSIGNMENT = CodeTemplate("""\
+${type} ${name} = ${type_cast}(node->${method}(stringToSymbol("${name}")));\
+""")
+
+CALL_NAMESPACE = CodeTemplate("at::${name}(${args})")
+CALL_METHOD = CodeTemplate("vars[0].${name}(${args})")
+
+CONSTRUCTOR = CodeTemplate("""\
+{"${descriptor}", [](Node *node) {
+  ${assignments}
+  return TensorOp([=](const variable_list& vars) -> variable_list {
+    return pack_list(${call});
+  }, "${name}", ${num_inputs});
+}},
+""")
+
+
+def is_jit_op(decl):
+    return (not decl['api_name'].endswith('_') and
+            not decl['name'].endswith('_out') and
+            not decl['name'].endswith('_forward') and
+            not any(arg['simple_type'] == 'Generator' for arg in decl['arguments']) and
+            not any(arg['simple_type'] == 'SparseTensor' for arg in decl['arguments']) and
+            not decl['return_type'] in FALLTHROUGH_RETURN_TYPES and
+            not decl['name'] in FALLTHROUGH_FUNCTIONS)
+
+
+def gen_jit_dispatch(declarations, out):
+    aten_decls = load_aten_declarations(declarations)
+    jit_decls = [d for d in aten_decls if is_jit_op(d)]
+
+    def is_tensor_arg(arg):
+        return arg['simple_type'] in {'Tensor', 'TensorList'}
+
+    ops = {}
+    for decl in jit_decls:
+        arguments = decl['arguments']
+        name = decl['name']
+        scalar_args = [arg for arg in arguments if not is_tensor_arg(arg)]
+
+        # Descriptor is a unique identified for a particular overload of an op
+        attr_names = sorted([arg['name'] for arg in scalar_args])
+        num_inputs = len(arguments) - len(scalar_args)
+        descriptor = '-'.join([decl['name'], str(num_inputs)] + attr_names)
+
+        # All scalar args need to be assigned, so they can be captured by a lambda
+        assignments = [ATTR_ASSIGNMENT.substitute(type=arg['simple_type'],
+                                                  type_cast=TYPE_CASTS.get(arg['simple_type'], ''),
+                                                  name=arg['name'],
+                                                  method=ATTR_METHOD_MAP[arg['simple_type']])
+                       for arg in scalar_args]
+
+        # Generate the actuall ATen call. This gets a bit tricky because of
+        # TensorList arguments, and functions that are only available as methods.
+        if 'namespace' in decl['method_of']:
+            if any(arg['simple_type'] == 'TensorList' for arg in arguments):
+                assert sum(map(is_tensor_arg, arguments)) == 1
+                args = ['as_tensor_list(vars)' if is_tensor_arg(arg) else arg['name']
+                        for arg in arguments]
+            else:
+                tensor_id = iter(count(start=0))
+                args = ['vars[{}]'.format(next(tensor_id)) if is_tensor_arg(arg) else arg['name']
+                        for arg in arguments]
+            call = CALL_NAMESPACE.substitute(locals())
```

Comment:
nit `locals()`:  this makes it really hard to figure out which things matter in the template. Maybe just list them verbosely.

Reply:
I know, it's just been really annoying for me that the only two other ways I could think of is either building up the env dict, which adds a ton of dict indexing into the code and makes it super verbose, or specifying 10 arguments to this function 😕